File: VB/Mvc_DbStorage_Sample_VB/Controllers/HomeController.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports DevExpress.DataAccess.Sql
Imports System.Web.Mvc
Imports DevExpress.Xpo
Imports Mvc_DbStorage_Sample_VB.Services.DAL

Namespace Mvc_DbStorage_Sample.Controllers
	Public Class HomeController
		Inherits Controller

		<HttpGet>
		Public Function Index() As ActionResult
			Using session = SessionFactory.Create()
				Dim reports = session.Query(Of ReportEntity)().Select(Function(x) New ReportModel With {.Url = x.Url}).ToArray()
				Dim firstReport = reports.FirstOrDefault()
				Dim model = New IndexModel With {
					.SelectedReportUrl = If(firstReport IsNot Nothing, firstReport.Url, String.Empty),
					.Reports = reports
				}
				Return View("Index", model)
			End Using
		End Function

		<HttpPost>
		Public Function Delete(ByVal url As String) As ActionResult
			Using session = SessionFactory.Create()
				Dim report = session.GetObjectByKey(Of ReportEntity)(url)
				session.Delete(report)

				session.CommitChanges()
			End Using
			Return Index()
		End Function

		<HttpPost>
		Public Function Design(ByVal url As String) As ActionResult
			Return View("Designer", New DesignModel With {
				.Url = url,
				.DataSource = CreateSqlDataSource()
			})
		End Function

		Private Function CreateSqlDataSource() As SqlDataSource
			Dim ds As New SqlDataSource("NWindConnectionString")
			ds.Name = "NWind"
			Dim categoriesQuery As SelectQuery = SelectQueryFluentBuilder.AddTable("Categories").SelectAllColumns().Build("Categories")
			ds.Queries.Add(categoriesQuery)
			ds.RebuildResultSchema()
			Return ds
		End Function

	End Class
End Namespace

File: CS/Mvc_DbStorage_Sample/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using DevExpress.DataAccess.Sql;
using Mvc_DbStorage_Sample.Services.DAL;
using System.Web.Mvc;
using DevExpress.Xpo;

namespace Mvc_DbStorage_Sample.Controllers {
    public class HomeController : Controller {
        [HttpGet]
        public ActionResult Index() {
            using (var session = SessionFactory.Create()) {
                var reports = session.Query<ReportEntity>()
                    .Select(x => new ReportModel {
                        Url = x.Url
                    })
                    .ToArray();
                var firstReport = reports.FirstOrDefault();
                var model = new IndexModel {
                    SelectedReportUrl = firstReport != null ? firstReport.Url : String.Empty,
                    Reports = reports
                };
                return View("Index", model);
            }
        }

        [HttpPost]
        public ActionResult Delete(string url) {
            using (var session = SessionFactory.Create()) {
                var report = session.GetObjectByKey<ReportEntity>(url);
                session.Delete(report);

                session.CommitChanges();
            }
            return Index();
        }

        [HttpPost]
        public ActionResult Design(string url) {
            return View("Designer", new DesignModel { Url = url, DataSource = CreateSqlDataSource() });
        }

        SqlDataSource CreateSqlDataSource() {
            SqlDataSource ds = new SqlDataSource("NWindConnectionString");
            ds.Name = "NWind";
            SelectQuery categoriesQuery = SelectQueryFluentBuilder.AddTable("Categories").SelectAllColumns().Build("Categories");
            ds.Queries.Add(categoriesQuery);
            ds.RebuildResultSchema();
            return ds;
        }

    }
}
